core/src/main/java/com/opensymphony/xwork2/ognl/SecurityMemberAccess.java

```diff
@@ -237,33 +242,27 @@ protected String toPackageName(Class<?> clazz) {
 
     protected boolean isExcludedPackageNamePatterns(Class<?> clazz) {
         String packageName = toPackageName(clazz);
-        for (Pattern pattern : excludedPackageNamePatterns) {
-            if (pattern.matcher(packageName).matches()) {
-                return true;
-            }
-        }
-        return false;
+        return excludedPackageNamePatterns.stream().anyMatch(pattern -> pattern.matcher(packageName).matches());
     }
 
     protected boolean isExcludedPackageNames(Class<?> clazz) {
-        String suffixedPackageName = toPackageName(clazz) + ".";
-        for (String excludedPackageName : excludedPackageNames) {
-            if (suffixedPackageName.startsWith(excludedPackageName)) {
+        String packageName = toPackageName(clazz);
+        List<String> packageParts = Arrays.asList(packageName.split("\\."));
+        for (int i = 0; i < packageParts.size(); i++) {
+            String parentPackage = String.join(".", packageParts.subList(0, i + 1));
+            if (excludedPackageNames.contains(parentPackage)) {
                 return true;
             }
         }
         return false;
     }
 
     protected boolean isClassExcluded(Class<?> clazz) {
-        if (clazz == Object.class || (clazz == Class.class && !allowStaticFieldAccess)) {
-            return true;
-        }
-        return excludedClasses.stream().anyMatch(clazz::isAssignableFrom);
+        return excludedClasses.contains(clazz);
```

Comment:
Constant time operation BUT it changes the semantics, more on this in my comment on `SecurityMemberAccessTest`